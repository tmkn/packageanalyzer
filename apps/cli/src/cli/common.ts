import { Command } from "clipanion";

import { AbstractReport } from "../../../../packages/shared/src/reports/Report.js";
import { ReportService } from "../../../../packages/shared/src/reports/ReportService.js";
import { Formatter, type IFormatter } from "../../../../packages/shared/src/utils/formatter.js";
import { NodeHost } from "../../../../packages/node/src/host/NodeHost.js";

export abstract class CliCommand<
    T extends AbstractReport<any> | AbstractReport<any>[]
> extends Command {
    abstract getReports(): T | Promise<T>;

    exitCode = 0;

    beforeProcess: ((report: T) => void) | undefined = undefined;

    async execute(): Promise<number | void> {
        const host = new NodeHost(this.context.stdout, this.context.stderr);

        try {
            const reports = await this.getReports();
            const reportService = new ReportService(
                {
                    reports: Array.isArray(reports) ? reports : [reports]
                },
                host
            );

            this.beforeProcess?.(reports);
            this.exitCode = (await reportService.process()) ?? 0;
        } catch (e: unknown) {
            const stderrFormatter: IFormatter = new Formatter(host.getStderrWriter());

            if (e) stderrFormatter.writeLine(e?.toString());
            this.exitCode = 1;
        } finally {
            await Promise.all([host.getStdoutWriter().flush(), host.getStderrWriter().flush()]);
        }

        return this.exitCode;
    }
}
