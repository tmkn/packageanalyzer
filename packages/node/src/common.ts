import * as path from "path";
import * as fs from "fs";
import { fileURLToPath } from "url";

import dayjs from "dayjs";

import type { DependencyTypes } from "../../shared/src/reports/Validation.js";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const __dirname = path.dirname(fileURLToPath(import.meta.url));
        // adjust when moving this function to another folder
        const file = path.join(__dirname, "..", "..", "..", "package.json");
        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}
