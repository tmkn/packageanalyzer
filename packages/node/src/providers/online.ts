import { QueryClient } from "@tanstack/query-core";

import {
    type IPackageMetadata,
    type IUnpublishedPackageMetadata
} from "../../../shared/src/npm.js";
import type { Url } from "../../../shared/src/reports/Validation.js";
import { AbstractPackageProvider } from "../../../shared/src/providers/provider.js";
import { downloadJson } from "../utils/requests.js";

//loads npm data from the web
export class OnlinePackageProvider extends AbstractPackageProvider {
    private readonly _queryClient = new QueryClient({
        defaultOptions: {
            queries: {
                retry: 3,
                retryDelay: 0, // with the default, the tests ran into timeouts, no delay yolo
                staleTime: Infinity // not really useful right now as caching is done via _cache
            }
        }
    });

    constructor(private readonly _url: Url) {
        super();
    }

    async getPackageMetadata(
        name: string
    ): Promise<IPackageMetadata | IUnpublishedPackageMetadata | undefined> {
        const cachedInfo = this._cache.get(name);

        if (typeof cachedInfo !== "undefined") {
            return cachedInfo;
        } else {
            const packageMetadata = await this._queryClient.fetchQuery({
                queryKey: ["package", name],
                queryFn: async ({ signal }) => {
                    const response = await downloadJson<
                        IPackageMetadata | IUnpublishedPackageMetadata
                    >(`${this._url}/${encodeURIComponent(name)}`, { signal });

                    if (response === null) {
                        console.error(
                            `Failed to fetch package metadata for "${name}" from ${this._url}, RETRYING...`
                        );
                        throw new Error(
                            `Failed to fetch package metadata for "${name}" from ${this._url}`
                        );
                    }

                    return response;
                }
            });

            return packageMetadata;
        }
    }
}

export const npmOnline = new OnlinePackageProvider(`https://registry.npmjs.com`);
