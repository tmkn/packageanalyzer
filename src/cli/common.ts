import * as path from "path";
import * as fs from "fs";

import * as dayjs from "dayjs";
import * as chalk from "chalk";

import { PackageAnalytics, VersionSummary, GroupedLicenseSummary } from "../analyzers/package";
import { DependencyTypes } from "../visitors/visitor";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const file = path.join(__dirname, "./../../package.json");

        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch (e) {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}

export function printStatistics(pa: PackageAnalytics): void {
    const padding = 40;
    const paddingLeft = 4;

    console.log(chalk.bold(`Statistics for ${pa.fullName}\n`));

    printPublished(pa, padding);
    printOldest(pa.oldest, padding);
    printNewest(pa.newest, padding);
    console.log(`${`Direct dependencies:`.padEnd(padding)}${pa.directDependencyCount}`);
    console.log(`${`Transitive dependencies:`.padEnd(padding)}${pa.transitiveDependenciesCount}`);
    printDistinctDependencies(pa.distinctByNameCount, pa.distinctByVersionCount, paddingLeft);
    printMostReferred(pa.mostReferred, padding);
    printMostDependencies(pa.mostDependencies, padding);
    printMostVersion(pa.mostVersions, padding, paddingLeft);
    printLoops(pa, padding, paddingLeft);
    printLicenseInfo(pa.licensesByGroup, paddingLeft);
}

function printNewest(newest: PackageAnalytics | undefined, padding: number): void {
    if (newest && newest.published) {
        console.log(
            `${`Newest package:`.padEnd(padding)}${
                newest.fullName
            } - ${newest.published.toUTCString()} ${daysAgo(newest.published)}`
        );
        console.log(`${`Newest package path:`.padEnd(padding)}${newest.pathString}`);
    }
}

function printOldest(oldest: PackageAnalytics | undefined, padding: number): void {
    if (oldest && oldest.published) {
        console.log(
            `${`Oldest package:`.padEnd(padding)}${
                oldest.fullName
            } - ${oldest.published.toUTCString()} ${daysAgo(oldest.published)}`
        );
        console.log(`${`Oldest package path:`.padEnd(padding)}${oldest.pathString}`);
    }
}

function printPublished(pa: PackageAnalytics, padding: number): void {
    if (!pa.published) return;

    console.log(
        `${`Published:`.padEnd(padding)}${pa.published.toUTCString()} ${daysAgo(pa.published)}`
    );
}

function printDistinctDependencies(
    byName: number,
    byNameAndVersion: number,
    paddingLeft: number
): void {
    console.log(`Distinct dependencies:`);
    console.log(`${``.padStart(paddingLeft)}${byName}: distinct name`);
    console.log(`${``.padStart(paddingLeft)}${byNameAndVersion}: distinct name and version`);
}

function printLoops(pa: PackageAnalytics, padding: number, paddingLeft: number): void {
    const { loops, loopPathMap, distinctLoopCount } = pa;

    console.log(`${`Loops:`.padEnd(padding)}${loops.length} (${distinctLoopCount} distinct)`);

    if (distinctLoopCount > 0) {
        const [first] = loops.map(l => l.pathString).sort();

        console.log(`    affected Packages: [${[...loopPathMap.keys()].join(", ")}]`);
        console.log(`    e.g. ${first}`);

        if (loops.length > 1)
            console.log(`${``.padStart(paddingLeft)}+${loops.length - 1} additional loops`);
    }
}

function printMostDependencies(pa: PackageAnalytics, padding: number): void {
    console.log(
        `${`Most direct dependencies:`.padEnd(padding)}"${pa.name}": ${pa.directDependencyCount}`
    );
}

function printMostReferred(arg: [string, number], padding: number): void {
    console.log(`${`Most referred package:`.padEnd(padding)}"${arg[0]}": ${arg[1]}x`);
}

function printMostVersion(mostVerions: VersionSummary, padding: number, paddingLeft: number): void {
    console.log(`${`Package(s) with most versions:`.padEnd(padding)}`);

    for (const [name, versions] of mostVerions) {
        console.log(`${``.padStart(paddingLeft)}${name} [${[...versions].join(", ")}]`);
    }
}

function printLicenseInfo(groupedLicenses: GroupedLicenseSummary, paddingLeft: number): void {
    console.log(`Licenses:`);
    for (const { license, names } of groupedLicenses) {
        const threshold = 4;
        const samples: string[] = names.slice(0, threshold);

        if (names.length >= threshold)
            console.log(
                `${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}, +${
                    names.length - threshold
                } more]`
            );
        else console.log(`${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}]`);
    }
}
