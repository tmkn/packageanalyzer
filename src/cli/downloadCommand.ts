import { Command, Option } from "clipanion";
import { OnlinePackageProvider } from "..";

import { DownloadReport, IDownloadParams } from "../reports/DownloadCountReport";
import { ReportService } from "../reports/ReportService";
import { Url } from "../utils/requests";

export class DownloadCommand extends Command {
    public package?: string = Option.String(`--package`, {
        description: `the package to retrieve the download count e.g. typescript@3.5.1`
    });

    static override usage = Command.Usage({
        description: `show the download count for a NPM package`,
        details: `
            This command will show show the download count for a NPM package.
        `,
        examples: [[`Show the download count for a NPM package`, `$0 loops --package typescript`]]
    });

    public static DownloadUrl?: Url;
    public static PackageProvider?: OnlinePackageProvider;

    static override paths = [[`downloads`]];

    async execute() {
        if (typeof this.package !== "undefined") {
            const params: IDownloadParams = {
                pkg: this.package,
                url: DownloadCommand.DownloadUrl
            };
            const downloadReport = new DownloadReport(params);
            downloadReport.provider = DownloadCommand.PackageProvider;

            const reportService = new ReportService(
                {
                    reports: [downloadReport]
                },
                this.context.stdout
            );

            await reportService.process();
        }
    }
}
