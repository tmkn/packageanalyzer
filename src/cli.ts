#!/usr/bin/env node

const pkg = require("./../../package.json");

import * as dayjs from "dayjs";

import { resolveFromFolder } from "./resolvers/folderResolver";
import { resolveFromName } from "./resolvers/nameResolver";
import { npmOnline } from "./providers/onlineProvider";
import {
    PackageAnalytics,
    LicenseSummary,
    VersionSummary,
    getNameAndVersion,
    groupPackagesByLicense
} from "./analyzer";

let commandFound = false;

process.argv.forEach((arg, i) => {
    if (arg === "-v" && !commandFound) {
        console.log(`Version ${pkg.version}`);
        commandFound = true;
    } else if (arg === "-h" && !commandFound) {
        showHelp();
        commandFound = true;
    } else if (arg === "-f" && !commandFound) {
        let folder = process.argv[i + 1];

        cliResolveFolder(folder);
        commandFound = true;
    } else if (arg === "-o" && !commandFound) {
        let name = process.argv[i + 1];

        cliResolveName(name);
        commandFound = true;
    }
});

if (!commandFound) {
    console.log(`No command found`);
    showHelp();
}

function showHelp(): void {
    console.log(`Package Analyzer ${pkg.version}`);
    console.log(`Options:`);
    console.log(
        `${`-f`.padEnd(10)} ${`Analyze a local folder`.padEnd(60)} e.g. "npa -f path/to/project"`
    );
    console.log(
        `${`-o`.padEnd(10)} ${`Analyze a package, version optional, default latest`.padEnd(
            60
        )} e.g. "npa -o typescript(@3.5.1)"`
    );
    console.log(`${`-v`.padEnd(10)} ${`Prints version`.padEnd(60)}`);
    console.log(`${`-h`.padEnd(10)} ${`Display help`.padEnd(60)}`);
}

async function cliResolveFolder(folder: string | undefined): Promise<void> {
    if (typeof folder === "undefined") {
        console.log(`Missing folder path\n`);
        showHelp();

        return;
    }

    try {
        let pa: PackageAnalytics = await resolveFromFolder(folder);

        printStatistics(pa);
    } catch (e) {
        //console.log(e);
    }
}

async function cliResolveName(pkgName: string | undefined): Promise<void> {
    if (typeof pkgName === "undefined") {
        console.log(`Missing package name\n`);
        showHelp();

        return;
    }

    try {
        const [name, version] = getNameAndVersion(pkgName);
        let pa: PackageAnalytics;

        if (typeof version === "undefined") pa = await resolveFromName(name, npmOnline);
        else pa = await resolveFromName([name, version], npmOnline);

        printStatistics(pa);
    } catch (e) {
        console.log(e);
    }
}

function printStatistics(pa: PackageAnalytics): void {
    const padding = 40;
    const paddingLeft = 4;

    console.log(`Statistics for ${pa.fullName}\n`);

    printPublished(pa, padding);
    printOldest(pa.oldest, padding);
    printNewest(pa.newest, padding);
    console.log(`${`Direct dependencies:`.padEnd(padding)}${pa.directDependencyCount}`);
    console.log(`${`Transitive dependencies:`.padEnd(padding)}${pa.transitiveDependenciesCount}`);
    printDistinctDependencies(pa.distinctByNameCount, pa.distinctByVersionCount, paddingLeft);
    printMostReferred(pa.mostReferred, padding);
    printMostDependencies(pa.mostDependencies, padding);
    printMostVersion(pa.mostVersions, padding, paddingLeft);
    printLoops(pa.loops, padding, paddingLeft);
    printLicenseInfo(pa.licenses, paddingLeft);
}

function printNewest(newest: PackageAnalytics | undefined, padding: number): void {
    if (newest && newest.published) {
        console.log(
            `${`Newest:`.padEnd(padding)}${
                newest.fullName
            } - ${newest.published.toUTCString()} ${daysAgo(newest.published)}`
        );
    }
}

function printOldest(oldest: PackageAnalytics | undefined, padding: number): void {
    if (oldest && oldest.published) {
        console.log(
            `${`Oldest:`.padEnd(padding)}${
                oldest.fullName
            } - ${oldest.published.toUTCString()} ${daysAgo(oldest.published)}`
        );
    }
}

function daysAgo(date: number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}

function printPublished(pa: PackageAnalytics, padding: number): void {
    if (!pa.published) return;

    console.log(
        `${`Published:`.padEnd(padding)}${pa.published.toUTCString()} ${daysAgo(pa.published)}`
    );
}

function printDistinctDependencies(
    byName: number,
    byNameAndVersion: number,
    paddingLeft: number
): void {
    console.log(`Distinct dependencies:`);
    console.log(`${``.padStart(paddingLeft)}${byName}: distinct name`);
    console.log(`${``.padStart(paddingLeft)}${byNameAndVersion}: distinct name and version`);
}

function printLoops(loops: PackageAnalytics[], padding: number, paddingLeft: number): void {
    console.log(`${`Loops:`.padEnd(padding)}${loops.length}`);

    if (loops.length > 0) {
        let [first] = loops.map(l => l.pathString).sort();

        console.log(`    e.g. ${first}`);

        if (loops.length > 1)
            console.log(`${``.padStart(paddingLeft)}+${loops.length - 1} additional loops`);
    }
}

function printMostDependencies(pa: PackageAnalytics, padding: number): void {
    console.log(
        `${`Most direct dependencies:`.padEnd(padding)}"${pa.name}": ${pa.directDependencyCount}`
    );
}

function printMostReferred(arg: [string, number], padding: number): void {
    console.log(`${`Most referred package:`.padEnd(padding)}"${arg[0]}": ${arg[1]}x`);
}

function printMostVersion(mostVerions: VersionSummary, padding: number, paddingLeft: number): void {
    console.log(`${`Package(s) with most versions:`.padEnd(padding)}`);

    for (const [name, versions] of mostVerions) {
        console.log(`${``.padStart(paddingLeft)}${name} [${[...versions].join(", ")}]`);
    }
}

function printLicenseInfo(allLicenses: LicenseSummary, paddingLeft: number): void {
    let summary = new Map<string, number>();
    let longestLine = 0;

    for (const [, licenses] of allLicenses) {
        for (const [, license] of licenses) {
            let specificLicense = summary.get(license);

            if (!specificLicense) {
                summary.set(license, 1);
            } else {
                summary.set(license, specificLicense + 1);
            }

            if (license.length > longestLine) longestLine = license.length;
        }
    }

    console.log(`Licenses:`);
    for (const { license, names } of groupPackagesByLicense(allLicenses)) {
        let threshold = 4;
        let samples: string[] = names.slice(0, threshold);

        if (names.length >= threshold)
            console.log(
                `${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}, +${names.length -
                    threshold} more]`
            );
        else console.log(`${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}]`);
    }

    /*let longestLine = 0;

    for(const [name, licenses] of allLicenses) {
        for(const [version, license] of licenses) {
            let fullName = `${name}@${version}`;

            if(fullName.length > longestLine)
                longestLine = fullName.length;
        }
    }

    for(const [name, licenses] of allLicenses) {
        for(const [version, license] of licenses) {
            console.log(`${`${name}@${version}`.padEnd(longestLine + 1)}${license}`.padStart(8));
        }
    }*/
}
