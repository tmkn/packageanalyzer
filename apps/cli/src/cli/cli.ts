import { Cli, Builtins } from "clipanion";

import { AnalyzeCommand } from "./analyzeCommand.js";
import { UpdateInfoCommand } from "./updateInfoCommand.js";
import { DownloadCommand } from "./downloadCommand.js";
import { LoopsCommand } from "./loopsCommand.js";
import { TreeCommand } from "./treeCommand.js";
import { LicenseCheckCommand } from "./licenseCommand.js";
import { ReportCommand } from "./reportCommand.js";
import { DependencyDumperCommand } from "./dependencyDumpCommand.js";
import { TestCommand } from "./testCommand.js";
import { DiffCommand } from "./diffCommand.js";
import { LintCommand } from "./lintCommand.js";
import { getVersion } from "../../../../packages/node/src/common.js";

export const cli = new Cli({
    binaryLabel: `packageanalyzer`,
    binaryName: `pkga`,
    binaryVersion: getVersion()
});

//standard commands
cli.register(ReportCommand);
cli.register(AnalyzeCommand);
cli.register(UpdateInfoCommand);
cli.register(DownloadCommand);
cli.register(LoopsCommand);
cli.register(TreeCommand);
cli.register(LicenseCheckCommand);
cli.register(DiffCommand);
cli.register(LintCommand);

//development niche commands
cli.register(DependencyDumperCommand);

cli.register(TestCommand);

//built in commands
cli.register(Builtins.HelpCommand);
cli.register(Builtins.VersionCommand);
