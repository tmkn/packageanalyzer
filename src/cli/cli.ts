import { Cli, Builtins } from "clipanion";

import { AnalyzeCommand } from "./analyzeCommand";
import { UpdateInfoCommand } from "./updateInfoCommand";
import { DownloadCommand } from "./downloadCommand";
import { LoopsCommand } from "./loopsCommand";
import { TreeCommand } from "./treeCommand";
import { NpmDumpCommand } from "./npmDumpCommand";
import { getVersion } from "./common";
import { NpmDumpLookupCreatorCommand } from "./npmLookupCreatorCommand";
import { LicenseCheckCommand } from "./licenseCommand";
import { ReportCommand } from "./reportCommand";
import { DependencyDumperCommand } from "./dependencyDumpCommand";

export const cli = new Cli({
    binaryLabel: `packageanalyzer`,
    binaryName: `pkga`,
    binaryVersion: getVersion()
});

//standard commands
cli.register(ReportCommand);
cli.register(AnalyzeCommand);
cli.register(UpdateInfoCommand);
cli.register(DownloadCommand);
cli.register(LoopsCommand);
cli.register(TreeCommand);
cli.register(LicenseCheckCommand);

//development niche commands
cli.register(NpmDumpCommand);
cli.register(NpmDumpLookupCreatorCommand);
cli.register(DependencyDumperCommand);

//built in commands
cli.register(Builtins.HelpCommand);
cli.register(Builtins.VersionCommand);