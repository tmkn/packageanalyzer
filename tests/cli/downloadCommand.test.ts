import { PassThrough } from "stream";

import { BaseContext } from "clipanion";

import { cli } from "../../src/cli/cli";
import { createMockDownloadServer, IMockServer, createMockNpmServer } from "../server";
import { DownloadCommand } from "../../src/cli/downloadCommand";
import { TestWritable } from "../common";
import { OnlinePackageProvider } from "../../src/providers/online";

describe(`Download Command`, () => {
    const stdout = new TestWritable();
    const mockContext: BaseContext = {
        stdin: process.stdin,
        stdout: stdout,
        stderr: new PassThrough()
    };

    let downloadServer: IMockServer;
    let npmServer: IMockServer;

    beforeAll(async () => {
        downloadServer = await createMockDownloadServer();
        npmServer = await createMockNpmServer();
    });

    test(`--package`, async () => {
        const command = cli.process([`downloads`, `--package`, `react`]);

        expect.assertions(1);
        command.context = mockContext;
        DownloadCommand.DownloadUrl = `http://localhost:${downloadServer.port}/`;
        DownloadCommand.PackageProvider = new OnlinePackageProvider(
            `http://localhost:${npmServer.port}`
        );

        await command.execute();

        expect(stdout.lines).toMatchSnapshot();
    });

    afterAll(async () => {
        await downloadServer.close();
        await npmServer.close();
    });
});
