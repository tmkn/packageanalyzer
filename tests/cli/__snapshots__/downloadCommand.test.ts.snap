// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Download Command --package 1`] = `
Array [
  "react: 8609192 Downloads",
  "",
]
`;
